### PR TITLE
Fix begin/end for valarray crashing on empty instance

### PR DESCRIPTION
**Pull request by Saalvage**
This issue also causes range-based for loops to break for empty valarrays.

I also added regression tests, which I hope are fine.

If there's anything wrong just tell me and I'll happily adjust!

### CODE REVIEW

File: stl/inc/valarray [hunk 1]
```diff
@@ -591,22 +603,22 @@ void swap(valarray<_Ty>& _Left, valarray<_Ty>& _Right) noexcept {
 
 template <class _Ty>
 _NODISCARD _Ty* begin(valarray<_Ty>& _Array) {
-    return &_Array[0];
```
Comment (miscco) [inline stl/inc/valarray:594]: I would prefer to move them inline, but I am not sure whether there are any ABI concerns rjere

**Review by miscco — APPROVED** [review]

File: stl/inc/valarray [hunk 2]
```diff
@@ -60,6 +60,18 @@ class valarray { // store array with various indexing options
 private:
```
Comment (CaseyCarter) [inline stl/inc/valarray:60]: (Pre-existing) friend declarations aren't accessible, so there's nothing here to make `private`:
```suggestion
public:
```

File: stl/inc/valarray [hunk 3]
```diff
@@ -60,6 +60,18 @@ class valarray { // store array with various indexing options
...
+    template <class _Ty2>
+    friend const _Ty2* begin(const valarray<_Ty2>& _Array);
+
+    template <class _Ty2>
+    friend _Ty2* end(valarray<_Ty2>& _Array);
+
+    template <class _Ty2>
+    friend const _Ty2* end(const valarray<_Ty2>& _Array);
+
 public:
```
Comment (CaseyCarter) [inline stl/inc/valarray:75]:
```suggestion
```

File: stl/inc/valarray (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/valarray:594]: We conventionally don't declare function templates `inline` (if I'm interpreting "move them inline" correctly) since (1) it isn't needed to satisfy ODR, and (2) the inliner almost ignores the presence of `inline` as a hint.

File: tests/tr1/tests/valarray/test.cpp [hunk 4]
```diff
@@ -340,5 +340,24 @@ void test_main() { // test basic workings of valarray definitions
...
     CHECK_INT(v0.min(), 2);
 
+    // test range-based for
+    v0 = Mycont();
+    for (char c : v0) {
+        (void) c;
+        CHECK(false);
+    }
+
+    v0 = Mycont("\1\1\1", 3);
```
Comment (CaseyCarter) [inline tests/tr1/tests/valarray/test.cpp:350]: I was about to ask where you got this horrible style from, and then I saw this is a pre-existing tr1 test. (No change requested, just despair at the horrors that appear in some of our ancient tests.)

**Review by CaseyCarter — APPROVED** [review]
I'm going to apply my suggestions which clean up pre-existing weirdness in `<valarray>`.

File: stl/inc/valarray (hunk 1 above)
Comment (Saalvage) [reply stl/inc/valarray:594]: I think they mean simply moving the definitions up into the class, which I think is indeed a bit cleaner. Although if I'm not wrong this also causes the function templates to be instantiated whenever the valarray class is, which I'm not sure is a concern?

Comment (CaseyCarter) [reply stl/inc/valarray:594]: Defining these in the class body (absent redeclarations at namespace scope) would make them hidden friends. This is better design and would have been a great idea in 2011 when `valarray`'s `begin` and `end` overloads were standardized. Today it would likely break code in the wild which uses `std::begin(meow)` / `std::end(meow)` for a valarray `meow`, and `valarray` is just not worth the trouble.

TLDR: Moving them into the class body gains us nothing since we'd need to redeclare them at namespace scope.

Comment (AlexGuteniev) [reply stl/inc/valarray:594]: >  (2) the inliner almost ignores the presence of `inline` as a hint

To be precise, for /Ob1 it doesn't ignore inline.

File: tests/tr1/tests/valarray/test.cpp (hunk 4 above)
Comment (StephanTLavavej) [reply tests/tr1/tests/valarray/test.cpp:350]: Agreed - we don't generally modify the legacy `tr1` test suite, but this seems like a reasonable place to add coverage and there isn't a better place in `tests/std` without adding a new test there. Also no change requested.

File: tests/tr1/tests/valarray/test.cpp [hunk 5]
```diff
@@ -340,5 +340,24 @@ void test_main() { // test basic workings of valarray definitions
     CHECK_INT(v0.max(), 4);
     CHECK_INT(v0.min(), 2);
 
+    // test range-based for
+    v0 = Mycont();
+    for (char c : v0) {
```
Comment (StephanTLavavej) [inline tests/tr1/tests/valarray/test.cpp:345]: No change requested: I generally recommend that all range-based for-loops should be written with `auto&`, `const auto&`, or (in rare cases where proxies are relevant) `auto&&`. (This avoids the potential harms of copying elements: losing modifications, taking the address of soon-to-be-destroyed objects, etc.) However, there are no actual correctness concerns here, and we don't imitate the `tr1` legacy test suite's style elsewhere, so I don't think it's worth changing this and resetting testing.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this bug - and congratulations on your first microsoft/STL commit! :tada: :heart_eyes_cat: :rocket:

This will ship as part of VS 2022 17.2 Preview 2.
